### PR TITLE
Adding listen address option so that IPv6 and loopback interfaces can be used.

### PR DESCRIPTION
**Pull request by grifball**
reference: #2477

### CODE REVIEW

**Comment by grifball** [comment]
I signed the cla. Not exactly sure how this works. Is that all I have to do?

**Comment by maksimr** [comment]
LGTM

**Comment by maksimr** [comment]
@grifball could you squash commits in to one

Thanks

**Comment by grifball** [comment]
I started writing something that automatically tries to set hostname when listenAddress is set but hostname isn't, but it seems out of scope. Figuring out which IP to use for hostname based on the value of listenAddress is really difficult, and can even be impossible in more complex interface setups. Maybe this isn't what you had in mind when you said "influence?"
I think that a better solution would be to give a warning if listenAddress is set and hostname isn't.

**Comment by grifball** [comment]
Let me know what you think of this solution. I'm not sure how to change the label on this PR.
I'll squash the commits once the code looks good.

File: lib/config.js [hunk 1]
```diff
@@ -199,6 +199,21 @@ var normalizeConfig = function (config, configFilePath) {
...
+  testConfig.hostname = null
+  testConfig.listenAddress = null
+  if (configModule != null) {
+    configModule(testConfig)
+  }
+  if (cliOptions != null) {
+    testConfig.set(cliOptions)
+  }
+  if (testConfig.hostname == null && testConfig.listenAddress != null) {
+    log.warn('ListenAddress is set but hostname isn\'t. If your browsers fail ' +
```
Comment (dignifiedquire) [inline lib/config.js]: It would be nice to print the current values of both the variables here.

File: lib/config.js [hunk 2]
```diff
@@ -199,6 +199,21 @@ var normalizeConfig = function (config, configFilePath) {
     throw new TypeError('Invalid configuration: formatError option must be a function.')
   }
 
+  // if the user changed listenAddress, but didn't set a hostname, warn them
+  var testConfig = new Config()
+  testConfig.hostname = null
+  testConfig.listenAddress = null
+  if (configModule != null) {
+    configModule(testConfig)
```
Comment (dignifiedquire) [inline lib/config.js]: Let's not do this, as this means we are evaluating the code from the user twice. Best case this slows things down, worst case it has unexpected side effects. You can add the check probably around the original invocition of the `configModule`

**Comment by dignifiedquire** [comment]
Thank you, yes I think it's fine to just warn the user about this. Left some comments about the implementation.

**Comment by grifball** [comment]
Alright, I implemented those changes. Let me know if anything else needs changing.

**Comment by wesleycho** [comment]
Do we want weak equality (`==` & `!=`) checks here? Seems like a better way would be to check specifically if it is a non-trivial string.

**Comment by grifball** [comment]
This should be a safe enough check. Unless they're setting these values to null and expecting the server to work correctly, which explicit string checks wouldn't solve anyway.
If they're setting these values to non-string types, I assume there's other checks in the config file for this, but it's out of scope in this PR anyway. The added code won't change how we handle config type checking.

**Comment by grifball** [comment]
I added a simple test to make sure that listen address was called with the configured listenAddress. That should cover all the functionality for this PR.

**Comment by dignifiedquire** [comment]
Thank you :octocat:

**Pull request merged**